### PR TITLE
Update pubnub and clear logs

### LINKED ISSUE #17: Try updating pubnub to v4
It obviously has breaking changes https://www.pubnub.com/docs/nodejs/migration-guide-sdk-v4#sdk-initialization

### PR DESCRIPTION
**Pull request by emirotin**
Closes #17 #18 

So code-wise I believe it's ready.
I've switched the tests from mocks (that I truly find useless here because they mask the _real_ API) to using the actual PubNub with random channel names.
And actually these tests just helped me find the bug in the code.

Now there are two problems with them:
1) they fail in the browser. I think because PubNub history does not guarantee we receive _everything_ published so far (even though I do wait on successful publish before trying to retrieve the history). So I've added some pretty random delays and they work in node but not in phantom.
2) tests fail on Windows (AppVeyor) and I don't know why

### CODE REVIEW

**Comment by emirotin** [comment]
The tests are unstable and new functionality (`clear`) is not covered yet, but the body of the code is OK to review

**Review by LucianBuzzo — APPROVED** [review]
LGTM

**Comment by emirotin** [comment]
Ah OK seems the windows failure is now the same browser failure (it used to be something different in the node part).

The browser failure is essentially history getting 0 results when it's supposed to get 3.
Will try playing with delays as I don't have any better ideas.

AH, and of course I've tested this by hand in the real UI and it works fine.

File: README.md [hunk 1]
```diff
@@ -34,7 +34,10 @@ Documentation
 
 * [logs](#module_logs)
     * [.subscribe(pubnubKeys, device)](#module_logs.subscribe) ⇒ <code>EventEmitter</code>
-    * [.history(pubnubKeys, device)](#module_logs.history) ⇒ <code>Promise.&lt;Array.&lt;Object&gt;&gt;</code>
+    * [.history(pubnubKeys, device, [options])](#module_logs.history) ⇒ <code>Promise.&lt;Array.&lt;Object&gt;&gt;</code>
+    * [.historyAfterClear(pubnubKeys, device, [options])](#module_logs.historyAfterClear) ⇒ <code>Promise.&lt;Array.&lt;Object&gt;&gt;</code>
```
Comment (pimterry) [inline README.md]: `historySinceLastClear` is a clearer method name for this imo (pun intended)

File: gulpfile.coffee [hunk 2]
```diff
@@ -15,13 +15,14 @@ OPTIONS =
 
 gulp.task 'coffee', ->
 	gulp.src(OPTIONS.files.app)
-		.pipe(coffee(bare: true)).on('error', gutil.log)
+		.pipe(coffee(bare: true, header: true)).on('error', gutil.log)
```
Comment (pimterry) [inline gulpfile.coffee:18]: :+1:

File: lib/logs.coffee [hunk 3]
```diff
@@ -59,23 +59,37 @@ utils = require('./utils')
...
+		if message.channel is clearChannel
+			return emit('clear')
+
+		if message.channel is channel
+			extractMessages(message.message).forEach (payload) ->
+				emit('line', payload)
+
+	pubnubListener = message: onMessage
+
+	instance.addListener(pubnubListener)
```
Comment (pimterry) [inline lib/logs.coffee]: Looks like we no longer subscribe to 'error' here. Is that right?

File: lib/utils.coffee [hunk 4]
```diff
@@ -16,20 +16,53 @@ limitations under the License.
...
-# @returns {String} logs channel
+# @returns {String} logs channel name
 #
 # @example
 # channel = utils.getChannel('...')
 ###
-exports.getChannel = (device) ->
-	return "device-#{device.logs_channel or device.uuid}-logs"
+exports.getChannel = (device, suffix = 'logs') ->
+	return "device-#{getBaseChannel(device)}-#{suffix}"
```
Comment (pimterry) [inline lib/utils.coffee]: We've duplicated this method - it's exactly the same as the method above.

File: tests/config.coffee [hunk 5]
```diff
@@ -0,0 +1,3 @@
+module.exports =
+	publishKey: 'pub-c-02e5848f-c9c5-4e6a-8586-268c44635766'
+	subscribeKey: 'sub-c-4a67af8a-049f-11e7-b09b-0619f8945a4f'
```
Comment (pimterry) [inline tests/config.coffee:3]: Is it definitely safe to publish these keys on github?

**Review by pimterry — CHANGES_REQUESTED** [review]

File: README.md (hunk 1 above)
Comment (emirotin) [reply README.md]: Right

File: lib/utils.coffee (hunk 4 above)
Comment (emirotin) [reply lib/utils.coffee]: Should be already fixed, noticed it myself

File: tests/config.coffee (hunk 5 above)
Comment (emirotin) [reply tests/config.coffee:3]: The keys used in production are anyway published to the entire world: https://api.resin.io/config

The channel name itself is here the _secret secret_.

File: lib/logs.coffee (hunk 3 above)
Comment (emirotin) [reply lib/logs.coffee]: Well, I forgot about it but the way to get it has changed. Will include it now

Comment (emirotin) [reply lib/logs.coffee]: Oh wow, there's no way to have it anymore I think. `subscribe` doesn't return a promise.

File: lib/utils.coffee (hunk 4 above)
Comment (emirotin) [reply lib/utils.coffee]: Oops wrong. What I've noticed was the duplicate `getBaseChannel`. Looks like I had some buggy local merge.

**Comment by emirotin** [comment]
Should I just disable the automated browser tests by excluding them from the default `test` script? They run locally and we also test it indirectly in the UI, but they fail consistently in the CI.

File: lib/logs.coffee (hunk 3 above)
Comment (pimterry) [reply lib/logs.coffee]: Really? If you're sure, we should drop it from the docs and examples.

**Review by pimterry — APPROVED** [review]

**Comment by pimterry** [comment]
I don't think we should disable the tests - we'll forget about them pretty quickly if we do. We should get them fixed, but that doesn't _have_ to block things. If we need this out in time to make the SDK 6.0 release, and we've tested this manually, then I think this is good to merge now, and we can come back and improve the tests later (and it'll be easier to remember to do that if they stay failing).

File: lib/logs.coffee (hunk 3 above)
Comment (emirotin) [reply lib/logs.coffee]: Try for yourself: https://www.pubnub.com/docs/nodejs-javascript/pubnub-javascript-sdk
I didn't find anything related there.

Comment (emirotin) [reply lib/logs.coffee]: Ah, found something: `PNNetworkIssuesCategory` event type can be used in the status callback

Comment (emirotin) [reply lib/logs.coffee]: Will add, though it's hard to test.

**Pull request merged**